doc/python-type-hinting.md

```diff
@@ -64,24 +62,100 @@ def test(self, var: Union[None, int, str]) -> None:
 
-## Forward declarations
+## Self Type
+
+To annotate methods that return an instance of their class, the `Self` type is used as it is bound to it's encapsulating class. In the example below, the type checker will correctly infer the type of `Circle().set_scale(0.5)` to be `Circle`:
+
+```python
+from typing import Self
+
+class Shape:
+    def set_scale(self, scale: float) -> Self:
+        self.scale = scale
+        return self
+
+
+class Circle(Shape):
+    def set_radius(self, radius: float) -> Self:
+        self.radius = radius
+        return self
+```
+
+This is recommended as forward declarations are now redundant in 3.10
+https://peps.python.org/pep-0673/
+
+## Type Alias
+
+Use the special annotation `TypeAlias` to declare type aliases more explicitly so type checkers are able to distinguish between type aliases and ordinary assignments:
+
+```python
+MyType: TypeAlias = "ClassName"
+def foo() -> MyType: ...
+```
+
+## Type Ignore
+
+To mark portions of the program that should not be covered by type hinting, use the following on a particular line:
+
+```python
+# type: ignore
+```
+
+A `# type: ignore` comment on a line by itself at the top of a file silences all errors in the file.
+
+`# type: ignore` should only be used when unavoidable. Ensure that a comment is added to explain why it has been used and have a prefix of `Type ignore:`.
+
+## ParamSpec
 
-To reference a type before it has been declared e.g. using a class as a type within the class declaration, add the special `annotations` import:
+Used to forward the parameter types of one callable to another callable. This is useful when type annotating decorators.
+
+For example, a basic logging function decorator can be typed as:
 
 ```python
-from __future__ import annotations
+T = TypeVar('T')
+P = ParamSpec('P')
+
+def add_logging(f: Callable[P, T]) -> Callable[P, T]:
+    '''A type-safe decorator to add logging to a function.'''
+    def inner(*args: P.args, **kwargs: P.kwargs) -> T:
+        logging.info(f'{f.__name__} was called')
+        return f(*args, **kwargs)
+    return inner
+
+@add_logging
+def add_two(x: float, y: float) -> float:
+    '''Add two numbers together.'''
+    return x + y
+```
 
-class TestClass:
+## TypeVar
 
-  def test(self, var: Sequence[TestClass]) -> None:
+Use `TypeVar` when the type returned by a function is the same as the type which was passed in i.e. the return type is linked to the parameter type:
+
+```python
+T = TypeVar('T')
+
+def increment_value(self, value: T) -> T:
+    return value + 1
+```
+
+`TypeVar` also accepts extra positional arguments to restrict the type parameter for improving code documentation and error prevention.
+
+```python
+T = TypeVar('T', int, float)
+
+def increment_value(self, value: T) -> T:
+    return value + 1
 ```
 
-This import is not necessary in Python 3.10.
+This tells the typechecker that values other than `int` and `float` are not allowed, resulting in an error if a string is passed for example:
```

Comment:
Think the examples are ok, maybe some of the sentences here don't add much over what is already in the docs and not really team specific?

Reply:
Sections refactored to be more team specific